cpp/tests/stream_compaction/distinct_count_tests.cpp

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2019-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/copying.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/stream_compaction.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/table_utilities.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <ctgmath>
+
+using cudf::nan_policy;
+using cudf::null_equality;
+using cudf::null_policy;
+
+template <typename T>
+struct DistinctCountCommon : public cudf::test::BaseFixture {
+};
+
+TYPED_TEST_SUITE(DistinctCountCommon, cudf::test::NumericTypes);
+
+TYPED_TEST(DistinctCountCommon, NoNull)
+{
+  using T = TypeParam;
+
+  auto const input = cudf::test::make_type_param_vector<T>(
+    {1, 3, 3, 4, 31, 1, 8, 2, 0, 4, 1, 4, 10, 40, 31, 42, 0, 42, 8, 5, 4});
+
+  cudf::test::fixed_width_column_wrapper<T> input_col(input.begin(), input.end());
+
+  cudf::size_type const expected = std::set<double>(input.begin(), input.end()).size();
```

Comment:
Does the type need to be `T` instead of `double`? This applies to a few other tests, too.
```suggestion
  cudf::size_type const expected = std::set<T>(input.begin(), input.end()).size();
```

Reply:
I think it's mainly to reduce build time. I had the same question and then found such explicit instantiation can reduce build time from 20 seconds to about 9 seconds.

Reply:
Let's add a comment to indicate that, since it's not very obvious and makes me question the correctness...